main/src/ca/uwaterloo/flix/language/phase/unification/RestrictableChooseInference.scala

```diff
@@ -0,0 +1,101 @@
+package ca.uwaterloo.flix.language.phase.unification
+
+import ca.uwaterloo.flix.api.Flix
+import ca.uwaterloo.flix.language.ast.{Ast, Kind, KindedAst, SourceLocation, Symbol, Type, TypeConstructor}
+import ca.uwaterloo.flix.language.phase.Typer
+import ca.uwaterloo.flix.language.phase.Typer.inferExp
+import ca.uwaterloo.flix.language.phase.unification.InferMonad.traverseM
+import ca.uwaterloo.flix.language.phase.unification.Unification.unifyTypeM
+import ca.uwaterloo.flix.util.InternalCompilerException
+import ca.uwaterloo.flix.util.collection.ListOps.unzip4
+
+object RestrictableChooseInference {
+
+  /**
+    * Returns the domain of the list of rules,
+    * i.e., the set of tags in the patterns.
+    */
+  private def dom(rules: List[KindedAst.RestrictableChoiceRule]): List[Symbol.RestrictableCaseSym] = {
+    rules.map {
+      case KindedAst.RestrictableChoiceRule(KindedAst.RestrictableChoicePattern.Tag(sym, _, _, _), _, _) => sym.sym
+    }
+  }
+
+  /**
+    * Converts the list of restrictable case symbols to a set type.
+    */
+  private def toType(syms: List[Symbol.RestrictableCaseSym], enumSym: Symbol.RestrictableEnumSym, loc: SourceLocation): Type = {
```

Comment:
If you have `n` cases do you get `n` case unions? Seems expensive...

Reply:
Yes. But we're not doing unification each time. It's just constructing the type {Case1} + {Case2} + {Case3} ...